File: 백준알고리즘/이분 탐색/javascript/공유기 설치.js
const fs = require('fs');
const stdin = fs.readFileSync('백준알고리즘/input.txt').toString().split('\n');

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

conditions = input().split(' ');
const N = +conditions[0];
const C = +conditions[1];

File: 백준알고리즘/이분 탐색/python/공유기 설치.py
import sys
input = sys.stdin.readline

N, C = map(int, input().split())
house = [int(input()) for _ in range(0, N)]

house.sort()


def count(d):
    cnt = 1
    ep = house[0]

    for i in range(1, N):
        if house[i] - ep >= d:
            cnt += 1
            ep = house[i]
    return cnt


def install_router():
    answer = 0
    sp = 1
    ep = house[-1] - house[0]

    while sp <= ep:
        md = (sp + ep) // 2
        if count(md) >= C:
            answer = md
            sp = md + 1
        else:
            ep = md - 1

    return answer


print(install_router())
